pkg/ccl/streamingccl/streamingest/stream_ingestion_frontier_processor.go

```diff
@@ -526,3 +533,50 @@ func (sf *streamIngestionFrontier) maybePersistFrontierEntries() error {
 	sf.lastFrontierDump = timeutil.Now()
 	return nil
 }
+
+func (sf *streamIngestionFrontier) maybeCheckForLaggingNodes() error {
+	defer func() {
+		sf.lastNodeLagCheck = timeutil.Now()
+	}()
+	checkFreq := streamingccl.ReplanFrequency.Get(&sf.FlowCtx.Cfg.Settings.SV)
+	maxLag := streamingccl.InterNodeLag.Get(&sf.FlowCtx.Cfg.Settings.SV)
+	if checkFreq == 0 || maxLag == 0 || timeutil.Since(sf.lastNodeLagCheck) < checkFreq {
+		return nil
+	}
+	ctx := sf.Ctx()
+	// Don't check for lagging nodes if the hwm has yet to advance.
+	if sf.replicatedTimeAtStart.Equal(sf.persistedReplicatedTime) {
+		log.VEventf(ctx, 2, "skipping lagging nodes check as hwm has yet to advance past %s", sf.replicatedTimeAtStart)
+		return nil
+	}
+	executionDetails, err := sf.gatherLaggingNodeInfo(ctx)
+	if err != nil {
+		log.Warningf(ctx, "could not gather information to check for lagging nodes: %s", err.Error())
+		return nil
+	}
+	return checkLaggingNodes(
+		executionDetails,
+		maxLag,
+	)
+}
+
+func (sf *streamIngestionFrontier) gatherLaggingNodeInfo(
+	ctx context.Context,
+) ([]frontierExecutionDetails, error) {
+	jobID := jobspb.JobID(sf.spec.JobID)
+
+	var specs []byte
+	if err := sf.FlowCtx.Cfg.DB.Txn(ctx, func(ctx context.Context, txn isql.Txn) error {
+		var err error
+		specs, err = jobs.ReadChunkedFileToJobInfo(ctx, replicationPartitionInfoFilename, txn, jobID)
```

Comment:
This is a very good point. I added a new field to the frontier processor proto and plumbed the partition specs from planning. It cleaned up the code nicely.

I'm not worried about back portability or mixed version stuff because:  a) this change is c2c only; b) it seems like we don't support mixed version c2c pre 23.2; c) the job coordinator, which creates the frontier spec, lives on the same node as the frontier spec, so they'll always be on the same version.